UI Component - Account

ulisesmac opened the pull request:
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

Fixes #14603

### Summary

[comment]: # (Summarise the problem and how the pull request solves it)

It implements the [Accounts Component](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System?node-id=2051%3A32614&t=5QOzYVaAMhY5MSPN-0) for both dark & white themes.

It solves it by adding the following namespaces:
- Styles: `quo2.components.settings.accounts.style`.
- Hiccup structure: `quo2.components.settings.accounts.view`.
- Preview-screen: `status-im2.contexts.quo-preview.settings.accounts`.

Additionally, 
- Adds the new component to `status-im2.contexts.quo-preview.main`.
- Solves an issue of the `account-avatar` component (in `quo2.components.avatars.account-avatar`) since this issue required it.

### Review notes
I splitted the code in three commits in order to make the code review easier, these commits do the following:

####  a0d0c5c02621b598a749453641df4fa0c73c4d5c  Fix account-avatar color & allow color customization 
The avatar component preview isn't working in the develop branch, it's due to wrongly extracting the `:purple` theme color from `quo2.foundations.colors/customization`, I fixed it by using `custom-color-by-theme`

Additionally, the avatar color was fixed to be `:purple`, but in this issue I needed to put it in different colors, so now `account-avatar` receives a `:color` key which defaults to `:purple` if not given.

#### a43794ba3f444f00fa7ea83776c992929632a0de Implement accounts component
Implements the Accounts component and adds it to `quo2.core`.

#### 4f1afcb86288c9fe1ec6ee2168532db9ed79a725 Create account-component preview
Creates a preview in the Settings section to test it.

### Testing notes
Navigate to `Profile -> Quo2.0 Preview -> Settings -> Accounts` in the app, switch the theme to test if the implementation meet the design.

- The menu button can be pressed, it prints in the terminal a string.
- A custom color, account name & account address can be set in the preview screen.

#### Platforms
- Android
- iOS

#### Areas that maybe impacted
<!-- (Optional. Specify if some specific areas has to be tested, for example 1-1 chats) -->

##### Functional
None, since it is only in a preview screen.

### Steps to test
Compile & run the app, login and naviage to  `Profile -> Quo2.0 Preview -> Settings -> Accounts`

<!-- (PRs will only be accepted if squashed into single commit.) -->

status: waiting for review.

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | b2f2ccce | [#1](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-14643/1/) | 2022-12-26 21:44:11 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/tests/job/PR-14643/1/consoleText) |

ulisesmac commented:
I need more details on what is expected.

Currently, as this is a first version, I created all the UI from zero, but I already know that existing components should be used.

Currently, the UI looks as follows:
![image](https://user-images.githubusercontent.com/90291778/209584877-7199783c-5f13-4f42-a096-ef6c0bf1770e.png)

In order to progress with the implementaition I need to solve some questions:
- What should be the namespace for this component?
- What properties (API) should this component receive?
- Where to use this component (who'll be calling it)?
- Do you alreaady have a three-dots-icon in the codebase? if so, where is it? if not, what should I do to add the icon?
- Currently, the implementation receives a boolean that indicates if the theme is dark or white, I know this is not the best way to achieve the theme change, could you please explain to me the way you are detecting the theme change or share me an example of how are you defining your theme-based components?
- What is the component that should be used for the  three-dot button? What will happen when pressed?
- What is the component that should be used for the avatar?
- Is localization important in this issue? (maybe not, since I guess everything shown in this UI is user info)
- In the figma model, the "Booze for Dubai" text has a weight of `600` but in Android, that style looks as `500`, temprarily I set it as `bold` (same as `700`) in order to look like the model, but it's not perfect. What should I do?
- Is this branch name correct (14643-ui-component-account)?

Thanks!

Parveshdhull commented:
Hi @ulisesmac, Thank you very much for contributing.
> What should be the namespace for this component?

All components namespace should follow the same hierarchy as the corresponding Figma design file, so it will be
`quo2.components.settings.accounts`

> What properties (API) should this component receive?

- Account name
- Account address
- Account avatar
- Customization color (For customizing card theme, for example in design it is blue)
- `on-press` callback for dots menu

> Where to use this component (who'll be calling it)?

For now, we should only provide a preview screen where we can test this component.
You can access preview screens for other components in `Profile -> Quo2.0 Preview` and the corresponding code for the preview should be in
`status-im2.contexts.quo-preview.settings.accounts`

> Do you alreaady have a three-dots-icon in the codebase? if so, where is it? if not, what should I do to add the icon?

three-dots-icon [example](https://github.com/status-im/status-mobile/blob/99f7788c22abb0a1933a943427a9e84a29c1c435/src/quo2/components/list_items/preview_list.cljs#L85) use, but we should use quo2/button instead (as mentioned below)

> Currently, the implementation receives a boolean that indicates if the theme is dark or white, I know this is not the best way to achieve the theme change, could you please explain to me the way you are detecting the theme change or share me an example of how are you defining your theme-based components?

We are using [theme-colors](https://github.com/status-im/status-mobile/blob/99f7788c22abb0a1933a943427a9e84a29c1c435/src/quo2/foundations/colors.cljs#L237) and [custom-color-by-theme](https://github.com/status-im/status-mobile/blob/99f7788c22abb0a1933a943427a9e84a29c1c435/src/quo2/foundations/colors.cljs#L215), you can check out [example](https://github.com/status-im/status-mobile/blob/99f7788c22abb0a1933a943427a9e84a29c1c435/src/quo2/components/buttons/dynamic_button.cljs#L12)

> What is the component that should be used for the three-dot button? What will happen when pressed?

For all new designs, we should use already implemented quo2 components only, so in this case, we can use the `quo2/button` component
(checkout button preview and [example](https://github.com/status-im/status-mobile/blob/99f7788c22abb0a1933a943427a9e84a29c1c435/src/status_im2/contexts/shell/cards/view.cljs#L152))

> What is the component that should be used for the avatar?

Account Avatar
Design - https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System?node-id=483%3A19401&t=I8EWnuwRbSh9j5PJ-0
Code - https://github.com/status-im/status-mobile/blob/develop/src/quo2/components/avatars/account_avatar.cljs
(Probably component is not ready to use, so might need some modification)

> Is localization important in this issue? (maybe not, since I guess everything shown in this UI is user info)

Please can you elaborate

> In the figma model, the "Booze for Dubai" text has a weight of 600 but in Android, that style looks as 500, temprarily I set it as bold (same as 700) in order to look like the model, but it's not perfect. What should I do?

Please use `quo2.text` and checkout `quo2.foundations.typography`, this has 600 weight.

> Is this branch name correct (14643-ui-component-account)?

Looks good to me

ulisesmac committed:
Fix account-avatar color & allow custom colors

ulisesmac committed:
Implement accounts component

ulisesmac committed:
Create account-component preview

ulisesmac commented:
Hi! @Parveshdhull 

>     Is localization important in this issue? (maybe not, since I guess everything shown in this UI is user info)

>> Please can you elaborate

Since I don't know how your localization system works, I was wondering if I'd need it to solve this issue, but realized that it's not necessary.

Your answers were really helpful! Thank you!

I'll update this PR description with the details of this finished PR.

**Note: I squashed this branch**

Parveshdhull commented:
Hi @ulisesmac, Is PR ready for review? If so, please can you remove WIP from the title?

ibrkhalil commented on src/quo2/components/settings/accounts/view.cljs:
```diff
@@ -0,0 +1,45 @@
+
+   [rn/view {:style style/card-bottom}
+    ;; Account name
+    [text/text {:size :paragraph-1, :weight :semi-bold}
```
Extra comma?

ulisesmac replied:
I'm using that comma as a visual separation when I write a map without a newline between the key-value pairs.I usually write:

```clojure
{:a 1
 :b 2}
;; or
{:a 1, :b 2}
```
If I'm not meeting the code-style in this project, I can remove those commas.

ilmotta replied:
FYI: Commas are removed automatically by the code formatter, if you have it enabled in your editor of course. `make lint-fix` will do the trick.

ulisesmac replied:
It didn't remove any of my commas (could it be a bug?), I manually remove them in f2e326b3092dca3724fceed710e3c608ee067b86

ibrkhalil commented on src/status_im2/contexts/quo_preview/settings/accounts.cljs:
```diff
@@ -0,0 +1,48 @@
+    :key     :custom-color
+    :type    :select
+    :options (mapv (fn [[k _]]
+                     {:key k, :value (string/capitalize (name k))})
```
I don't see comma's as white space to be necessary, but it's up to you

ilmotta replied:
As @ulisesmac did, commas can help with readability, but the current rules used in `.zprintrc` will remove them anyway.

ulisesmac replied:
Removed in f2e326b3092dca3724fceed710e3c608ee067b86

ibrkhalil commented on src/quo2/components/settings/accounts/view.cljs:
```diff
@@ -0,0 +1,45 @@
+   [rn/view {:style (style/background-top custom-color)}]
+   [rn/view {:style (style/background-bottom)}]])
+
+(defn avatar [{:keys [_color _icon] :as avatar-props}]
```
Why destructure here if we don't need the keys ?

ulisesmac replied:
I use those vars as some kind of code documentation.

If we only pass `avatar-props` it could be a little harder to know what keys this function needs to work.

That's the reason why I added them, should I remove them?

ilmotta replied:
It's an interesting approach @ulisesmac, but it can quickly get outdated due to the lack of usage of the bindings. In essence, the function `avatar` doesn't care about what's inside `avatar-props`, as long as it's a map, it just needs to assoc a default `:size`, so to me it's the opposite, not having the destructuring of the map helps me realize what's inside the map is not important.

ulisesmac replied:
I think it's a matter of style and preferences, I understand your point of view. I removed the destructured parameters in ac8644a301daad136ded2f3290379f8b20b611ae.

ilmotta replied:
It really is, although I argued in favor of removing them, it's subjective.

ibrkhalil commented on src/quo2/components/settings/accounts/view.cljs:
```diff
@@ -0,0 +1,45 @@
+  [rn/view {:style style/card}
+   [card-background {:custom-color custom-color}]
+   [rn/view {:style style/card-top}
+    ;; Avatar
```
I think **most** of these these comments are unnecessary, As the vars they annotate have the same textual content

ulisesmac replied:
I agree,
I deleted all of these comments, since it's clear what the UI is defining below them.

I added 57acac2ed0769bbf7f8936a600e9c6327a703ace & 425e7ae51662824569fb1f53faa5ceecf8e11f07 with the changes.

ulisesmac commented:
> Hi @ulisesmac, Is PR ready for review? If so, please can you remove WIP from the title?

Yes, It's ready. I updated the title, thanks!

ulisesmac committed:
Delete unnecessary comments

ulisesmac committed:
Delete unnecessary comments

ilmotta commented on src/quo2/components/avatars/account_avatar.cljs:
```diff
@@ -27,12 +22,11 @@
-                               (:light icon-color-value))
+  [{:keys [size icon color]
+    :or   {size  80
+           icon  :main-icons/placeholder
```
The keyword qualification should be just `:i`. https://github.com/status-im/status-mobile/blob/develop/doc/new-guidelines.md#icons

Please, let us know if you think anything could be improved in the guidelines.

ulisesmac replied:
I was not aware of those guidelines, thanks for sharing!

I replaced it in 415173139a52d171e755945da1f3018ca7b111cc

This use of  `:main-icons` was already in develop (I didn't added it), Must we always replace those appearances?

ilmotta replied:
You're right about that, I actually thought we recently eliminated all references to the old icon qualification.

I understand it's not totally obvious from the outside, but we've been trying to establish some stronger conventions in the code, such as the icon keyword qualification. Right now coding in the Status app is a bit of a minefield in terms of knowing what's old or not (esp. for non-core collaborators), but we're trying to document the new guidelines as much as possible. Feedback is highly appreciated :)

ilmotta commented on src/status_im2/contexts/quo_preview/settings/accounts.cljs:
```diff
@@ -0,0 +1,48 @@
+                       ;; Fixed parameters:
+                       :avatar-icon     :main-icons/placeholder
+                       :on-press-menu   (fn []
+                                          (prn "menu pressed"))})]
```
For fake event handler feedback, we've been using `(js/alert ...`), so designers using the preview also get feedback if something happened.

ulisesmac replied:
It sounds good!
Solved in 5d66eb63d5418f2660063535946350b7d96ca840

ilmotta commented on src/status_im2/contexts/quo_preview/settings/accounts.cljs:
```diff
@@ -0,0 +1,48 @@
+  (let [state (r/atom {:custom-color    :blue
+                       :account-name    "Booze for Dubai"
+                       :account-address "0x21a ... 49e"
+                       ;; Fixed parameters:
```
This comment seems unnecessary to me.

ulisesmac replied:
Removed in 889cab4e3d8cee5fdb584cb9e0c9d6810e714d05

ilmotta commented on src/status_im2/contexts/quo_preview/settings/accounts.cljs:
```diff
@@ -0,0 +1,48 @@
+                       :account-name    "Booze for Dubai"
+                       :account-address "0x21a ... 49e"
+                       ;; Fixed parameters:
+                       :avatar-icon     :main-icons/placeholder
```
The keyword qualification should be just :i. https://github.com/status-im/status-mobile/blob/develop/doc/new-guidelines.md#icons

ulisesmac replied:
Solved in 415173139a52d171e755945da1f3018ca7b111cc

ulisesmac commented:
@Parveshdhull 
I'm sorry, accidentally removed the reviewers: @OmarBasem @ilmotta & you whe I clicked in the "reload" icon in the reviewers zone.

Could you please request them to review again?

ilmotta commented on src/status_im2/contexts/quo_preview/settings/accounts.cljs:
```diff
@@ -0,0 +1,48 @@
+(ns status-im2.contexts.quo-preview.settings.accounts
+  (:require [clojure.string :as string]
+            [quo2.components.settings.accounts.view :as quo2]
```
You can simply alias quo2.core as `quo` so that the code is already future proof when we eliminate the old quo component library.

ulisesmac replied:
Ok ok, I changed it in 46b87118c8aad9bf9f8fcc7c6fb9acf546cf59eb

ilmotta commented on src/status_im2/contexts/quo_preview/settings/accounts.cljs:
```diff
@@ -0,0 +1,48 @@
+  [{:label   "Custom color"
+    :key     :custom-color
+    :type    :select
+    :options (mapv (fn [[k _]]
```
Unless there's a strict need for the collection to be a vector, we prefer to keep it as a lazy sequence, so just use `map`.

ulisesmac replied:
OK! Solved in 496d54c1fd7d4d80b320f45666e39f5682e90a14

ilmotta reviewed (commented):
@ulisesmac, could you call `make lint-fix` to make sure everything is properly formatted?

ilmotta commented on src/quo2/components/settings/accounts/view.cljs:
```diff
@@ -0,0 +1,40 @@
+(ns quo2.components.settings.accounts.view
```
@ulisesmac, we have also started to *attempt* to write "component tests" for new components in `quo2/`. It would be very welcomed if you could test some things.

You can find examples in `__tests__` directories. Documentation about testing components is on the way, but right now the best is to just stick to the existing examples in the codebase.

ulisesmac replied:
Ok, I'll take a look at those examples and think in some tests. I'll reply when done

ilmotta replied:
Feel free to share any problems you might find. `make component-test-watch` should be enough to get you going.

- Example test: `src/quo2/components/selectors/filter/component_spec.cljs`
- You need to add the new test namespace to `src/quo2/core_spec.cljs`.

ulisesmac replied:
@ilmotta I've already had a look at those tests, but I'm not clear how to write them, also I'm not familiar with `testing-library/react-native` yet.
So I'd like to wait for the documentation on how to write & run tests in order to write my own.

Could this MR be accepted even if it doesn't have tests?

ilmotta replied:
Understandable @ulisesmac, we're not blocking PRs lacking such tests. There's definitely a learning curve to write them effectively.

ulisesmac committed:
Remove `:main-icons` ns-keyword

ulisesmac committed:
Set js/alert instead of prn

ulisesmac committed:
Remove unnecessary comment

ulisesmac committed:
Change quo2 to quo

ulisesmac committed:
Keep seq lazy

ulisesmac committed:
Remove commas

ulisesmac committed:
Remove unused destructured params

ulisesmac commented:
> @ulisesmac, could you call `make lint-fix` to make sure everything is properly formatted?

As told you before, I ran `make lint-fix` and didn't change any of my code (it kept my commas), I ran it at the root of the project, am I doing something wrong?

ilmotta commented:
> > @ulisesmac, could you call `make lint-fix` to make sure everything is properly formatted?
> 
> As told you before, I ran `make lint-fix` and didn't change any of my code (it kept my commas), I ran it at the root of the project, am I doing something wrong?

@ulisesmac, I believe this has been fixed just today. You opened a PR right about the time we're evolving how the code is formatted using zprint :man_shrugging: See if rebasing against upstream solves the issue for you (it should), if not, please report to us.

ulisesmac committed:
Run make lint-fix

ulisesmac commented:
> > > @ulisesmac, could you call `make lint-fix` to make sure everything is properly formatted?
> > 
> > 
> > As told you before, I ran `make lint-fix` and didn't change any of my code (it kept my commas), I ran it at the root of the project, am I doing something wrong?
> 
> @ulisesmac, I believe this has been fixed just today. You opened a PR right about the time we're evolving how the code is formatted using zprint man_shrugging See if rebasing against upstream solves the issue for you (it should), if not, please report to us.

OK! I rebased my branch and ran the lint, I added a new commit with my code formatted.
Thanks!

ulisesmac commented:
> ![image](https://user-images.githubusercontent.com/33176106/209865156-bd82813a-0097-479f-9dad-c87577347d2a.png) Signing time!!

I rebased the branch on order to sign all the commits

flexsurfer commented on src/quo2/components/settings/accounts/view.cljs:
```diff
@@ -0,0 +1,46 @@
+(ns quo2.components.settings.accounts.view
+  (:require [quo.react-native :as rn]
```
react-native.core :as rn

ulisesmac replied:
Solved in 3d9a6bf0afe3430696defb1e18fa8de2d9a94eaa

flexsurfer commented on src/quo2/components/settings/accounts/view.cljs:
```diff
@@ -0,0 +1,46 @@
+(ns quo2.components.settings.accounts.view
+  (:require [quo.react-native :as rn]
+            [quo2.components.avatars.account-avatar :as av]
```
better to use fill name alias `[quo2.components.avatars.account-avatar :as account-avatar ]`

ulisesmac replied:
Totally agree!
Solved in 3d9a6bf0afe3430696defb1e18fa8de2d9a94eaa

flexsurfer commented on src/status_im2/contexts/quo_preview/settings/accounts.cljs:
```diff
@@ -0,0 +1,51 @@
+            [quo2.components.settings.accounts.view :as quo]
+            [quo2.foundations.colors :as colors]
+            [react-native.core :as rn]
+            [reagent.core :as r]
```
`[reagent.core :as reagent]`

ulisesmac replied:
Again, solved in 3d9a6bf0afe3430696defb1e18fa8de2d9a94eaa

ulisesmac committed:
Fix requires
